include/oneapi/dpl/internal/function.h

```diff
@@ -90,6 +91,123 @@ get_access(const Policy& policy, T* ptr)
     assert(sycl::get_pointer_type(ptr, policy.queue().get_context()) == sycl::usm::alloc::shared);
     return ptr;
 }
+
+template <typename ValueType, typename Policy, typename Iterator>
+ValueType
+get_data_at(const Policy& p, Iterator i, size_t index)
+{
+    return get_access<sycl::access::mode::read>(p, i)[index];
+}
+
+template <typename ValueType, typename Policy, typename T>
+ValueType
+get_data_at(const Policy& policy, T* ptr, size_t index)
+{
+    sycl::queue q = policy.queue();
+
+    switch (sycl::get_pointer_type(ptr, q.get_context()))
+    {
+    case sycl::usm::alloc::host:
+    case sycl::usm::alloc::shared:
+        return ptr[index];
+
+    case sycl::usm::alloc::device:
+    {
+        ValueType host_data;
+        q.copy(ptr + index, &host_data, 1);
+        q.wait();
+        return host_data;
+    }
+    break;
+
+    case sycl::usm::alloc::unknown:
+        assert(!"Unknown pointer type");
+    }
+
+    return {};
+}
+
+template <typename ValueType, typename Policy, typename Iterator>
+void
+set_data_at(const Policy& p, Iterator i, size_t index, ValueType val)
+{
+    get_access<sycl::access::mode::write>(p, i)[index] = val;
+}
+
+template <typename ValueType, typename Policy, typename T>
+void
+set_data_at(const Policy& policy, T* ptr, size_t index, ValueType val)
+{
+    sycl::queue q = policy.queue();
+
+    switch (sycl::get_pointer_type(ptr, q.get_context()))
+    {
+    case sycl::usm::alloc::host:
+    case sycl::usm::alloc::shared:
+        ptr[index] = val;
+        break;
+
+    case sycl::usm::alloc::device:
+        q.copy(&val, ptr + index, 1);
+        q.wait();
+        break;
+
+    case sycl::usm::alloc::unknown:
+        assert(!"Unknown pointer type");
+    }
+}
+
+template <typename Policy, typename IteratorSrc, typename IteratorDest>
+void
+copy_data_to(const Policy& p, IteratorSrc itSrc, size_t indexSrc, IteratorDest itDest, size_t indexDest)
+{
+    auto accRead = get_access<sycl::access::mode::read>(p, itSrc);
```

Comment:
Lines 164-165: `get_access` will cause the whole buffer to be transferred to the host, while we only need a single item. I think it could be improved.

Reply:
Fixed, please check.

Reply:
Fixed. Please review get_buffer_for and it's usage.